### PR TITLE
JAVA-2700: Revisit execution profile parameters in mapper

### PR DESCRIPTION
**Pull request by olim7t**
This amends JAVA-2633 (5e1e5107c91c3546e5cc81a29cbccdbbaff0f82b).

See also [this review](https://github.com/datastax/java-driver/pull/1409#pullrequestreview-377027405).

### CODE REVIEW

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java [hunk 1]
```diff
@@ -146,13 +146,6 @@ protected void populateBuilderWithFunction(
     }
   }
 
-  protected void populateBuilderWithProfile(CodeBlock.Builder builder) {
-    builder.beginControlFlow("if(context.getExecutionProfileName() != null)");
-    builder.addStatement(
-        "boundStatementBuilder = boundStatementBuilder.setExecutionProfileName(context.getExecutionProfileName())");
-    builder.endControlFlow();
-  }
-
```
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java:155]: One more thing I hadn't caught previously: it's easier to do this in `DaoBase`.

File: mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/DaoBase.java [hunk 2]
```diff
@@ -60,6 +60,11 @@
 
   protected static CompletionStage<PreparedStatement> prepare(
       SimpleStatement statement, MapperContext context) {
+    if (context.getExecutionProfileName() != null) {
+      statement = statement.setExecutionProfileName(context.getExecutionProfileName());
+    } else if (context.getExecutionProfile() != null) {
+      statement = statement.setExecutionProfile(context.getExecutionProfile());
+    }
     return context.getSession().prepareAsync(statement);
```
Comment (tomekl007) [inline mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/DaoBase.java:68]: Do we prevent users from setting both `profileName` and profile (`class`)?
If no, did we document the fact that `getExecutionProfileName` has priority over `getExecutionProfile`?

File: mapper-runtime/revapi.json [hunk 3]
```diff
@@ -60,8 +60,13 @@
       {
         "code": "java.method.addedToInterface",
         "new": "method java.lang.String com.datastax.oss.driver.api.mapper.MapperContext::getExecutionProfileName()",
-        "justification": "JAVA-2633 Adding execution profile to mapper"
+        "justification": "JAVA-2633: Add execution profile argument to DAO factory method (accept API break -- it's unlikely that MapperContext will be implemented outside of the driver)"
```
Comment (tomekl007) [inline mapper-runtime/revapi.json:63]: Can't we use `customState` to propagate this? We would not need to break backward compatibility in such a case

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DaoProfile.java [hunk 4]
```diff
@@ -26,18 +27,16 @@
  *
  * <p>Example:
  *
- * <p>*
- *
  * <pre>
  *  * &#64;Mapper
  *  * public interface InventoryMapper {
  *  *   ProductDao productDao(@DaoTable String executionProfile);
```
Comment (adutra) [inline mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DaoProfile.java]: Shouldn't this be `ProductDao productDao(@DaoProfile String executionProfile);`?

**Review by adutra — CHANGES_REQUESTED** [review]

File: mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/DaoBase.java (hunk 2 above)
Comment (olim7t) [reply mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/DaoBase.java:68]: It's enforced implicitly by the mapper processor: only one `@DaoProfile`-annotated parameter is allowed per DAO factory method.

I'll add a comment on the context methods to explain that.

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DaoProfile.java (hunk 4 above)
Comment (olim7t) [reply mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DaoProfile.java]: 👍 Good catch, will fix

File: mapper-runtime/revapi.json (hunk 3 above)
Comment (olim7t) [reply mapper-runtime/revapi.json:63]: Already discussed [on the original PR](https://github.com/datastax/java-driver/pull/1409#discussion_r389558576).

To expand: `MapperContext` can't be implemented by the user in any useful way (they can't inject their implementation in mapper-generated code), so we deemed it acceptable to break the API.

File: mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/DaoBase.java (hunk 2 above)
Comment (olim7t) [reply mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/DaoBase.java:68]: Now that I think of it, we could also allow injection of the profile via `MapperBuilder`, like we do with the default keyspace. I'll add that.

File: integration-tests/src/test/java/com/datastax/oss/driver/mapper/ProfileIT.java [hunk 5]
```diff
@@ -35,37 +36,55 @@
...
 import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.ClassRule;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TestRule;
 
 @Category(ParallelizableTests.class)
 public class ProfileIT {
```
Comment (olim7t) [inline integration-tests/src/test/java/com/datastax/oss/driver/mapper/ProfileIT.java:67]: I've completely rearranged this test to cover all combinations of mapper-level, dao-level, and method-level (via the existing statement attributes mechanism).

File: mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/DefaultMapperContext.java [hunk 6]
```diff
@@ -70,11 +56,6 @@ public DefaultMapperContext(
         NullAllowingImmutableMap.copyOf(customState));
   }
 
-  public DefaultMapperContext(
-      @NonNull CqlSession session, @NonNull Map<Object, Object> customState) {
-    this(session, null, customState);
-  }
-
```
Comment (olim7t) [inline mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/DefaultMapperContext.java:73]: This constructor wasn't used anymore, no need to keep it.

File: core/src/main/java/com/datastax/oss/driver/api/core/cql/StatementBuilder.java [hunk 7]
```diff
@@ -91,6 +91,7 @@ protected StatementBuilder(StatementT template) {
   @NonNull
   public SelfT setExecutionProfileName(@Nullable String executionProfileName) {
     this.executionProfileName = executionProfileName;
+    this.executionProfile = null;
     return self;
   }
```
Comment (olim7t) [inline core/src/main/java/com/datastax/oss/driver/api/core/cql/StatementBuilder.java:98]: This solved a bug when combining DAO-level profile and `@StatementAttributes`, but I think it also makes sense for general usage of statement builders.

**Review by tomekl007 — APPROVED** [review]

**Review by adutra — APPROVED** [review]

**Pull request merged**